### PR TITLE
Modify loopy to use four nested loops

### PR DESCRIPTION
**Pull request by topepo**
Prior ot this PR, there were a few helpers to recognize several cases: 
 
- Predict only: the model has no postprocessing at all
- Postprocessing with:
	- tuning and no estimation (i.e. calibration)
	- tuning and estimation 
	- no tuning and no estimation
	- no tuning and estimation 
 
 These lead to some efficiencies since it 

- Avoids refitting of tailors
- Efficiently uses `multi_predict()` for submodels
 
 and so on. 
 
 This PR removes these and adds additional loops over the prediction and postprocessing steps to: 
 
 - Cleaner code
 - Fewer helper functions
 - Easier code for object extraction

The old version was archived via a [release](https://github.com/topepo/melodie/releases/tag/v1.2.1.9000) so that we can easily access or install it if needed.

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -57,13 +60,16 @@ Suggests:
...
     tidymodels/hardhat,
     tidymodels/parsnip,
+    tidymodels/probably,
     tidymodels/recipes,
     tidymodels/rsample,
     tidymodels/tailor,
+    tidymodels/tune,
     tidymodels/workflows,
-    tidymodels/dials
+    KlausVigo/kknn#28
```
Comment (topepo) [inline DESCRIPTION:72]: Until it is merged and/or a new version is on CRAN

File: R/loopy-helpers.R [hunk 2]
```diff
@@ -1,4 +1,7 @@
 # Helpers for loopy()
+#' @import rlang
```
Comment (topepo) [inline R/loopy-helpers.R:2]: To make the standalones work

File: R/loopy-helpers.R [hunk 3]
```diff
@@ -175,55 +182,80 @@ has_tailor_estimated <- function(x) {
...
+  post_obj
+}
+
+# ------------------------------------------------------------------------------
+
+predict_all_types <- function(
+    wflow_fit,
+    static,
+    submodel_grid = NULL,
+    predictee = "assessment"
```
Comment (topepo) [inline R/loopy-helpers.R:212]: This governs which data are used to train and assess. Helps with inner splits

File: R/loopy-helpers.R [hunk 4]
```diff
@@ -638,3 +493,31 @@ reorder_pred_cols <- function(x, y_name) {
       .before = dplyr::everything()
     )
 }
+
+engine_to_parsnip <- function(wflow, grid) {
```
Comment (topepo) [inline R/loopy-helpers.R:497]: We have something like this in parsnip. tune _probably_ has something to do that but I could not find it. I'll add an issue to reassess whether we need it or not.

File: R/loopy.R [hunk 5]
```diff
@@ -55,12 +55,18 @@ loopy <- function(resamples, grid, static) {
     # --------------------------------------------------------------------------
     # Iterate over model parameters
 
+    # Make a copy of the current workflow so that we can finalize it multiple
+    # times
+    pre_wflow <- current_wflow
```
Comment (topepo) [inline R/loopy.R:60]: I figured out that the `finalize_*()` functions will not update parameters whose values currently are `tune()`. We save this to maintain those values.

File: R/loopy.R [hunk 6]
```diff
@@ -70,36 +76,105 @@ loopy <- function(resamples, grid, static) {
...
+        current_prd <- current_model$predict_stage[[1]][prd, ]
+
+        if (has_submodel) {
+          sub_nm <- get_sub_param(current_prd)
+          sub_grid <- current_prd[, sub_nm]
+
+          # The assigned submodel parameter (from min_grid()) is in the
+          # current grid. Remove that and add the one that we are predicting on
+
+          current_grid <- current_grid |>
```
Comment (topepo) [inline R/loopy.R:100]: Predicts one submodel at a time. We _could_ use `multi_predict()` to get those all at once and loop over that but this is more straightforward. We can benchmark if we are worried about there being a difference.

File: R/loopy.R [hunk 7]
```diff
@@ -70,36 +76,105 @@ loopy <- function(resamples, grid, static) {
...
+        for (pst in seq_len(num_pst_iter)) {
+          # cli::cli_inform("-- Postprocessing {pst} of {num_pst_iter}")
+
+          if (has_post) {
+            current_pst <- current_prd$post_stage[[1]][pst, ]
+            post_grid <- current_pst
+
+            current_post_grid <- rebind_grid(current_predict_grid, current_pst)
+
+            # make data for prediction (TODO maybe make a function)
```
Comment (topepo) [inline R/loopy.R:129]: A helper would make this cleaner

File: R/loopy.R [hunk 8]
```diff
@@ -70,36 +76,105 @@ loopy <- function(resamples, grid, static) {
...
+          }
+
+          # --------------------------------------------------------------------
+          # Allocate predictions to an overall object
+
+          pred_iter <- pred_iter + 1
+          # cli::cli_inform("-- -- Allocation {pred_iter} of {nrow(grid)}")
+          # TODO We might not be able to predict ahead of time how many rows should
+          # be in the reserve
+          # pred_reserve <- update_reserve(pred_reserve, pred_iter, final_pred, nrow(grid))
```
Comment (topepo) [inline R/loopy.R:163]: I punted on using the reserve for now. The initial cache size depends on the inner split (if any) so the logic needs to be updated.

File: inst/regression_tests/simple_tune_example.R [hunk 9]
```diff
@@ -0,0 +1,56 @@
+# ------------------------------------------------------------------------------
+# Expected results produced by tune 1.3.0 to compare to later implementations
+# This "simple" example is for model-only results with no submodels
+
+library(tidymodels)
```
Comment (topepo) [inline inst/regression_tests/simple_tune_example.R:5]: Are there any more that we should add?

File: tests/testthat/test-loopy-post-estimation-no-tuning.R [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+  submodel_only_res <- melodie:::loopy(rs_iter, submodel_only_grid, static_1)
+  expect_true(!is.null(submodel_only_res$.metrics[[1]]))
+  expect_named(submodel_only_res, c(".metrics", ".notes", "id", ".predictions"))
+  expect_true(nrow(submodel_only_res) == 1)
+  expect_equal(
+    nrow(submodel_only_res$.predictions[[1]]),
+    nrow(data_1$pred$data) * nrow(submodel_only_grid)
+  )
+
+  # TODO more tests can be added when calibration method = "none" is implemented
```
Comment (topepo) [inline tests/testthat/test-loopy-post-estimation-no-tuning.R:209]: We need to have a 1:1 unit test here. When tidymodels/tailor#75 we can update this

File: R/loopy.R [hunk 11]
```diff
@@ -55,12 +55,18 @@ loopy <- function(resamples, grid, static) {
     # --------------------------------------------------------------------------
     # Iterate over model parameters
 
+    # Make a copy of the current workflow so that we can finalize it multiple
+    # times
```
Comment (EmilHvitfeldt) [inline R/loopy.R:59]:
```suggestion
    # times, since finalize_*() functions will not update parameters whose 
    # values currently are tune()
```

File: R/loopy-helpers.R [hunk 12]
```diff
@@ -125,6 +128,10 @@ has_mod_param <- function(x) {
   any(names(x) != "predict_stage")
 }
 
+has_post_param <- function(x) {
+  nrow(x[[1]]) > 0
+}
+
```
Comment (EmilHvitfeldt) [inline R/loopy-helpers.R:134]: This function is not being used

File: R/loopy-helpers.R [hunk 13]
```diff
@@ -175,55 +182,80 @@ has_tailor_estimated <- function(x) {
...
+# ------------------------------------------------------------------------------
+
+predict_all_types <- function(
+    wflow_fit,
+    static,
+    submodel_grid = NULL,
+    predictee = "assessment"
+) {
+  predictee <- rlang::arg_match(predictee, c("assessment", "calibration"))
+  outputs <- get_output_columns(wflow_fit, syms = TRUE)
```
Comment (EmilHvitfeldt) [inline R/loopy-helpers.R:215]: Looking at `get_output_columns()` i noticed that the `syms` argument doesn't do anything.

I'll open an issue for it

File: R/loopy-helpers.R [hunk 14]
```diff
@@ -175,55 +182,80 @@ has_tailor_estimated <- function(x) {
...
+
+  if (predictee == "calibration" & static$post_estimation) {
```
Comment (EmilHvitfeldt) [inline R/loopy-helpers.R:217]:
```suggestion
  if (predictee == "calibration" && static$post_estimation) {
```

**Review by EmilHvitfeldt — APPROVED** [review]
looks good. it is more loops, but it makes for a very clean reading experience

File: tests/testthat/test-loopy-logging.R [hunk 15]
```diff
@@ -75,6 +75,7 @@ test_that("model error doesn't stop grid", {
 })
 
 test_that("prediction error doesn't stop grid", {
+  skip("Get Emil to take a look at this")
```
Comment (topepo) [inline tests/testthat/test-loopy-logging.R:78]: @EmilHvitfeldt I forgot to highlight this; it breaks during my testing and I don't know where the bug is (or if there is one)

**Review by hfrick — COMMENTED** [review]
I'm very supportive of the strategy to first write out the unoptimized quadruple loop, get the tests right, and then potentially adding optimizations back in.

`pre_update_fit()` and `model_update_fit()` update the workflow by splicing in the parameter candidates and then fitting it, so update and then fit, rather than updating the fit. Since there is already `.fit_[pre|model]()` in workflows, I'd like the idea of adding `.finalize_*()` functions and keeping the call to `.fit_*()` visible in the loop. To a certain extend, this also applies to `train_post()`. Is there a `.fit_post()` method in workflows already that would train the post-processor?

Regarding the naming:
- The schedule is my/a anchor through the four nested loops, so I'd love to be able to follow that object and its parts for the "current" stage more easily by adding "schedule" to the name, so e.g., `current_schedule_pre` instead of `current_pre`.
- The `static` object does make the interface a lot simpler - which is really helpful! But I'm nominating that for a round of thinking-about-the-name as well, because maybe we can hint at what's in the object in addition to the important info of "this doesn't change over the iterations".
- There are a few occasions where there are several versions of a thing and we generally have a pattern of going from general to specific so I'd rename, e.g., `num_pre_iter` and `num_mod_iter` to `num_iter_*` (or `iter_num_*` if we use the `iter` prefix more).
- We could also take naming inspiration from `tune:::tune_grid_loop_iter()` where applicable.

**Comment by topepo** [comment]
> I'd like the idea of adding `.finalize_*()` functions and keeping the call to `.fit_*()` visible in the loop

I'd rather not do this. First, we want the code to have more modular functions so that it doesn't devolve into a long script-like format. Second, it would be better to keep the two steps in a function so that we can invoke the catching/logging function once. 

Renaming is fine with me though. How about `{pre,mod}_finalize_fit()`?

**Comment by topepo** [comment]
For these other suggestions: 

> Regarding the naming

they sound great. Would you like to do the renaming? If so, do you want to do it here or merge then make a different PR?

**Comment by hfrick** [comment]
> How about {pre,mod}_finalize_fit()?

I'd move the pre/mod part to the end because when they get called, it's pretty clear that we are in the preprocessing or model loop, so "finalize and fit" is the more valuable piece of information. Plus, it follows the general-to-specific pattern we often use. Can I convince you? 😄 

> Would you like to do the renaming? If so, do you want to do it here or merge then make a different PR?

I'm happy to do the renaming. I'll do it in an additional PR after this one is merged.

**Comment by topepo** [comment]
> Can I convince you?

You have! 😄  

Go ahead and edit this PR for that change and then start another for the more extensive ones. Thanks!

**Comment by hfrick** [comment]
Is `train_post()` -> `finalize_fit_post()` also in scope?

**Comment by topepo** [comment]
> Is `train_post()` -> `finalize_fit_post()` also in scope? 

Yep! 

We you are done with your changes, please go ahead and merge.
